### PR TITLE
Refactored the bot class

### LINKED ISSUE #717: Unmanaged Resource leak
Ok, after learning a bit about unmanaged resource, I know that there is a lack of PROPER disposing of unmanaged resource(Example: File logging of SteamBot) which isn't good. @BlueRaja @scholtzm What you guys think?

### PR DESCRIPTION
**Pull request by WildCard65**
also closed #717 until modular logging is merged(which also fixes the issue)

### CODE REVIEW

**Comment by BlueRaja** [comment]
Are there other backwards-compatibility breaking changes?

File: SteamBot/Bot.cs [hunk 1]
```diff
@@ -17,109 +17,130 @@
...
+                MyInventoryTask.Wait();
+                return MyInventoryTask.Result;
             }
         }
 
-        private BackgroundWorker backgroundWorker;
+        /// <summary>
+        /// Compatibility sanity.
+        /// </summary>
+        [Obsolete("Refactered to be Log instead of log")]
```
Comment (scholtzm) [inline SteamBot/Bot.cs]: Refactored* :smile:

**Comment by WildCard65** [comment]
@scholtzm @FrozenHaxor calling you to find anything that breaks backwards compat in this(so I can restore and obselete)

**Comment by WildCard65** [comment]
@BlueRaja Is this mergeable now?

**Comment by WildCard65** [comment]
commit 9bb78e7 in this PR idk where it came from but it changed nothing major(only includes spellfix)

**Comment by scholtzm** [comment]
Should have probably used rebase rather then merge.

**Comment by WildCard65** [comment]
@BlueRaja still waiting for your input on mergability.

**Comment by BlueRaja** [comment]
I still don't understand why you renamed all the private variables from camelCase to PascalCase.  camelCase is the correct convention.

**Comment by scholtzm** [comment]
I agree with the refactoring itself, `Bot` class is very messy. However, @BlueRaja is right, private fields should be camelCased (or _camelCased but we don't use that convention in SteamBot).

**Comment by WildCard65** [comment]
Happy now @BlueRaja?

**Comment by BlueRaja** [comment]
Yes, with the formatting.  However, when I now run the code, the bot does not work.  It stalls after

    [(System) 2015-03-31 19:32:39] DEBUG: Starting bot with username BlueRajaBot.
    Type help for bot manager commands.
    botmgr >

File: SteamBot/Bot.cs [hunk 2]
```diff
@@ -233,16 +254,12 @@ public bool StartBot()
...
-            log.Info("Connecting...");
-
-            if (!backgroundWorker.IsBusy)
-                // background worker is not running
-                backgroundWorker.RunWorkerAsync();
-
+            Log.Info("Connecting...");
+            if (botThread.IsBusy)
+                return true; //We will consider this running.
+            botThread.RunWorkerAsync();
```
Comment (BlueRaja) [inline SteamBot/Bot.cs]: It looks like this change in logic is the cause of the bot not starting up.

File: SteamBot/Bot.cs [hunk 3]
```diff
@@ -17,45 +17,103 @@
...
+        public delegate UserHandler UserHandlerCreator(Bot bot, SteamID id);
+        #endregion
 
-        // The response to all chat messages sent to it.
-        public string ChatResponse;
+        #region Private readonly variables
+        private readonly SteamUser.LogOnDetails logOnDetails;
+        private readonly string schemaLang;
+        private readonly string logFile;
+        private readonly Dictionary<SteamID, UserHandler> userHandlers;
```
Comment (BlueRaja) [inline SteamBot/Bot.cs]: The old code was

    Dictionary<ulong, UserHandler> userHandlers = new Dictionary<ulong, UserHandler>();

`userHandlers` is now always null.

File: SteamBot/Bot.cs [hunk 4]
```diff
@@ -905,41 +901,41 @@ private void BackgroundWorkerOnRunWorkerCompleted(object sender, RunWorkerComple
...
             ICallbackMsg msg;
 
-            while (!backgroundWorker.CancellationPending)
+            while (!botThread.CancellationPending)
             {
                 try
                 {
                     msg = SteamClient.WaitForCallback(true);
-                    HandleSteamMessage(msg);
+                    HandleSteamMessage(msg); //Making handling of steam messages prevents bot from freezing on action(if userhandlers gets infinite loop).
```
Comment (BlueRaja) [inline SteamBot/Bot.cs]: This comment can be deleted.

**Comment by BlueRaja** [comment]
After making the above changes, the refactoring appears to work correctly.

File: SteamBot/Bot.cs (hunk 2 above)
Comment (WildCard65) [inline SteamBot/Bot.cs]: Do i leave this the same and apply the below change? Or change this as well?

Comment (BlueRaja) [inline SteamBot/Bot.cs]: You have to change this as well.  I changed it back to

    if(!botThread.IsBusy)
        backgroundWorker.RunWorkerAsync();

With the `return true` as it is now, `SteamClient.Connect()` is never called..

File: SteamBot/Bot.cs (hunk 3 above)
Comment (WildCard65) [inline SteamBot/Bot.cs]: Initialized in bot constructor now.

**Comment by scholtzm** [comment]
Formatting is eff-ed up -> https://github.com/Jessecar96/SteamBot/blob/master/SteamBot/Bot.cs#L500-L550

**Comment by WildCard65** [comment]
Your welcome to submit a pr to fix it.

**Pull request merged**